harness/determined/common/api/_session.py

```diff
@@ -95,13 +48,59 @@ class BaseSession(metaclass=abc.ABCMeta):
     a few cases, you might be ok with an UnauthSession.  BaseSession is really only to express that
     you don't know what kind of session you need.  For example, the generated bindings take a
     BaseSession because the protos aren't annotated with which endpoints are authenticated.
+
+    `BaseSession` and subclasses can be used directly, or as a context manager. When used as a
+    context manager, all requests within the context will share a persistent underlying HTTP
+    connection. When used directly, each request will create a new HTTP connection.
+
+    Example of direct usage:
+
+    .. code:: python
+
+       session = api.Session(...)
+       # Each request opens and closes a new HTTP connection.
+       session.get(...)
+       session.get(...)
+
+    Example of use as a context manager:
+
+    .. code:: python
+       with api.Session(...) as session:
+           # Each request within the context re-uses the same HTTP connection.
+           session.get(...)
+           session.get(...)
+
     """
 
     master: str
     cert: Optional[certs.Cert]
     _max_retries: Optional[GeneralizedRetry]
+    # Persistent HTTP session
+    _http_session: Optional[requests.Session]
+
+    def __enter__(self):
+        self.persist_http_session()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def persist_http_session(self) -> None:
```

Comment:
If this is meant to be a public method (or at least public within the repo), please give these docstrings.

Reply:
shouldn't be public. made it private.